### ISSUE TITLE
Spot instance is created, agent is not

### ISSUE THREAD

**Issue by jeanblanchard** [jenkinsci/ec2-fleet-plugin#163]
For a while now, we have been having an issue where the plugin never seems to create the jenkins agents, even though the spot instances are created correctly.

For the case where I attached the logs (see below), we have 1 ec2 instance, but the widget still says "State: modifying, label: "ec2-fleet", nodes: 0, target: 0"

It used to work. I'm not sure what broke it, system update, jenkins update, plugin update or otherwise. I tried downgrading the plugin up to 1.14 with no changes. Tried various plugin configurations, including with or without "No Delay Provision Strategy".

I'm not sure exactly what information to provide to find the issue. There does not seem to be any relevant logs... Is there a way to get the logs for the agent creation itself ?

Log extract attached for com.amazon.jenkins.ec2fleet : [com.amazon.jenkins.ec2fleet.txt](https://github.com/jenkinsci/ec2-fleet-plugin/files/4020197/com.amazon.jenkins.ec2fleet.txt)

Environment:
* Jenkins 2.204.1
* ec-fleet 1.17.0 (broken since at least 1.16.0)

**Comment by terma** [comment]
Based on log attached (thx a lot) looks like plugin stuck https://github.com/jenkinsci/ec2-fleet-plugin/blob/master/src/main/java/com/amazon/jenkins/ec2fleet/EC2FleetCloud.java#L496 probably connection problem, do you use customized endpoint? Do you see any exception in log after this time some http timeout in 5-10 min?

**Comment by jeanblanchard** [comment]
Thanks for looking into this.

No customized AWS endpoint, just using plain-old eu-west-1 region.

No logs at all about nodes or executors in this time period, and no exceptions.

Also I forgot to mention, after I disabled the plugin, I could manually create ssh agents on the ec2 instances created by the plugin, so the issue is most likely not an ssh error.

**Comment by terma** [comment]
I think I know what happens. First small explanation; EC2 Fleet Plugin has thread to sync EC2 Spot Fleet state with Jenkins, this thread run sync by default each ```10 sec``` and describe list of EC2 instances by Spot Request ID first then describe them by EC2 Instance ID to get actual state.

I thin in your case ```describeInstaces``` just failed by some reason permissions etc and because of code bug [here](https://github.com/jenkinsci/ec2-fleet-plugin/blob/master/src/main/java/com/amazon/jenkins/ec2fleet/EC2Api.java#L108) go to infinite loop, so it will never describe instance and proceed.

Let me try to fix this infinite loop, and you will be able see actual AWS API error.

**Comment by terma** [comment]
ok, fixed in version [1.17.1](https://github.com/jenkinsci/ec2-fleet-plugin/releases/tag/ec2-fleet-1.17.1) thx for report, now you should be able to see in Jenkins log why plugin cannot describe instances.

**Comment by jeanblanchard** [comment]
I can indeed now see the error:
```
Error during fleet FleetCloud stats update
com.amazonaws.services.ec2.model.AmazonEC2Exception: You are not authorized to perform this operation. (Service: AmazonEC2; Status Code: 403; Error Code: UnauthorizedOperation; Request ID: xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.handleErrorResponse(AmazonHttpClient.java:1712)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeOneRequest(AmazonHttpClient.java:1367)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeHelper(AmazonHttpClient.java:1113)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.doExecute(AmazonHttpClient.java:770)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeWithTimer(AmazonHttpClient.java:744)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.execute(AmazonHttpClient.java:726)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.access$500(AmazonHttpClient.java:686)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutionBuilderImpl.execute(AmazonHttpClient.java:668)
	at com.amazonaws.http.AmazonHttpClient.execute(AmazonHttpClient.java:532)
	at com.amazonaws.http.AmazonHttpClient.execute(AmazonHttpClient.java:512)
	at com.amazonaws.services.ec2.AmazonEC2Client.doInvoke(AmazonEC2Client.java:24586)
	at com.amazonaws.services.ec2.AmazonEC2Client.invoke(AmazonEC2Client.java:24553)
	at com.amazonaws.services.ec2.AmazonEC2Client.invoke(AmazonEC2Client.java:24542)
	at com.amazonaws.services.ec2.AmazonEC2Client.executeDescribeInstances(AmazonEC2Client.java:11753)
	at com.amazonaws.services.ec2.AmazonEC2Client.describeInstances(AmazonEC2Client.java:11724)
	at com.amazon.jenkins.ec2fleet.EC2Api.describeInstancesBatch(EC2Api.java:93)
	at com.amazon.jenkins.ec2fleet.EC2Api.describeInstances(EC2Api.java:72)
	at com.amazon.jenkins.ec2fleet.EC2Api.describeInstances(EC2Api.java:62)
	at com.amazon.jenkins.ec2fleet.EC2FleetCloud.updateByState(EC2FleetCloud.java:496)
	at com.amazon.jenkins.ec2fleet.EC2FleetCloud.update(EC2FleetCloud.java:428)
	at com.amazon.jenkins.ec2fleet.CloudNanny.doRun(CloudNanny.java:58)
	at hudson.triggers.SafeTimerTask.run(SafeTimerTask.java:70)
	at jenkins.security.ImpersonatingScheduledExecutorService$1.run(ImpersonatingScheduledExecutorService.java:58)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```
After updating the IAM policy (I had a restriction on the instance tags for `ec2:DescribeInstances`), I now get the log
```
FleetCloud [ec2-fleet] start
FleetCloud [ec2-fleet] fleet instances [i-xxxx]
FleetCloud [ec2-fleet] described instances [i-xxxx]
FleetCloud [ec2-fleet] jenkins nodes [i-xxxx]
```
And the node is correctly detected \o/

Thanks @terma handling this so quickly.
